docs/spec/concepts.rst

```diff
@@ -3,97 +3,216 @@
-
-Support for singleton types in unions
-^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
-
-A singleton instance is frequently used to mark some special condition,
-in particular in situations where ``None`` is also a valid value
-for a variable. Example::
-
-  _empty = object()
-
-  def func(x=_empty):
-      if x is _empty:  # default argument value
-          return 0
-      elif x is None:  # argument was provided and it's None
-          return 1
-      else:
-          return x * 2
-
-To allow precise typing in such situations, the user should use
-a union type in conjunction with the ``enum.Enum`` class provided
-by the standard library, so that type errors can be caught statically::
-
-  from enum import Enum
-
-  class Empty(Enum):
-      token = 0
-  _empty = Empty.token
-
-  def func(x: int | None | Empty = _empty) -> int:
-
-      boom = x * 42  # This fails type check
-
-      if x is _empty:
-          return 0
-      elif x is None:
-          return 1
-      else:  # At this point typechecker knows that x can only have type int
-          return x * 2
-
-Since the subclasses of ``Enum`` cannot be further subclassed,
-the type of variable ``x`` can be statically inferred in all branches
-of the above example. The same approach is applicable if more than one
-singleton object is needed: one can use an enumeration that has more than
-one value::
-
-  class Reason(Enum):
-      timeout = 1
-      error = 2
-
-  def process(response: str | Reason = '') -> str:
-      if response is Reason.timeout:
-          return 'TIMEOUT'
-      elif response is Reason.error:
-          return 'ERROR'
-      else:
-          # response can be only str, all other possible values exhausted
-          return 'PROCESSED: ' + response
+Static, dynamic, and gradual typing
+-----------------------------------
+
+A **statically-typed** programming language runs a type-checker before running
+a program. The program is required to be well-typed according to the language's
+type system. The type system assigns a type to all expressions in the language
+and verifies that their uses obey the typing rules. Normally, a program that is
+not well-typed (i.e., one that contains a type error) will not run. Java and
+C++ are examples of statically-typed object-oriented languages.
+
+A **dynamically-typed** programming language does not run a type-checker before
+running a program. Instead, it will check the types of values before performing
+operations on them at runtime. This is not to say that the language is
+"untyped". Values at runtime have a type and their uses obey typing rules. Not
+every operation will be checked, but certain primitive operations in the
+language such as attribute access or arithmetic will be. Python was
+historically a dynamically-typed language.
+
+**Gradual typing** is the name for a specific way to combine static and dynamic
+typing. Type-annotated Python allows mixing static and dynamic checking at a
```

Comment:
I'm not sure I see why "mix" is confusing. Despite the fact that dynamic type checking always occurs, once we also add some static type checking, there is a mix. Some parts of the program are now (also) statically checked, others rely solely on the dynamic checking.

But I will try to re-word to avoid the phrasing.